### PR TITLE
ENG-244 Create tests for `RStarM.sol`

### PR DESCRIPTION
**Pull request by 0xmovses**
Adds solidity unit tests to test the gorth16 verification capabilities of `RStarM.sol`

Tests are: 
`testRegisterValidator`
`testVerifyKnownGoodReceipt`
`testVerifyKnownGoodImageIdAndJournal`
`testVerifyMangledReceipts`
`testHonestValidators`
`testDishonestValidators`

### Run the tests 
`cd protocol-units/dispute && forge test`

https://movement.atlassian.net/browse/ENG-189?atlOrigin=eyJpIjoiNWU3ODE4OTQ4NGY2NDZlYWIxZDc2MGU4MTRlNDQzYzgiLCJwIjoiaiJ9

### CODE REVIEW

File: protocol-units/dispute/src/RStarM.sol
Comment (l-monninger) [inline protocol-units/dispute/src/RStarM.sol:240]: This is good. We may change this to the staked rollup concept sooner than later.

File: protocol-units/dispute/src/RStarM.sol
Comment (l-monninger) [inline protocol-units/dispute/src/RStarM.sol:196]: So, still for RStarM we would want those optimistic commitment rounds. But, hold off on patching that in for a bit.

File: protocol-units/dispute/src/RStarM.sol
Comment (l-monninger) [inline protocol-units/dispute/src/RStarM.sol:156]: It makes even more sense if we going with the staked rollup approach to call these `stake` and `unstake`.

File: protocol-units/dispute/test/RStarM.t.sol
Comment (l-monninger) [inline protocol-units/dispute/test/RStarM.t.sol:93]: I'd like to see some tests on the optimistic commitment. Include cases for fraudulent actors, honest actors, a mix, etc.

File: protocol-units/dispute/src/RStarM.sol
Comment (l-monninger) [inline protocol-units/dispute/src/RStarM.sol:248]: This is actually a race right now instead of m of n. You would have that the subsequent committers have to  match the commitment of the first committer.

Instead, you'll want to track all commitments and check each time whether you have m of n agreeing on a state.

Even after the state is agreed upon, you'll want to commit to track other commitments for slashing.

**Comment by l-monninger** [comment]
1. Tests are passing.
2. Rewards are omitted, but that's okay for now.
3. The last logical issue above needs to be resolved and tested. 
4. I'm going to push an update to the nix flake to include the foundry packages. 

Once 3 and 4 are done, we can merge this. We can address the rest in subsequent PRs.

**Review by l-monninger — CHANGES_REQUESTED** [review]

**Comment by l-monninger** [comment]
In the future, it would also be good to roll these up into some e2e tests. For now, it could just be something similar to the e2e tests with Sovereign Labs SDK from before.

**Comment by 0xmovses** [comment]
Two new tests have been added `honestValidators`, `dishonestValidators`. The race condition has been fixed also.

**Comment by l-monninger** [comment]
@0xmovses 

There's still confusion here: https://github.com/movementlabsxyz/movement/blob/6c3f0095febb72e382aea6615ad5dd421add6162/protocol-units/dispute/src/RStarM.sol#L250

1. Rounds are not global. There are rounds for a given block height. 
2. The race condition still exists. This is roughly what this should look like: https://github.com/movementlabsxyz/rfcs/blob/rfc/mcr/0026-r*m/rfc-r*m.md

You want to keep track of all of the different block commitments submitted. I'll update alongside MCR to show you what I'm talking about.

File: protocol-units/dispute/src/RStarM.sol
Comment (l-monninger) [inline protocol-units/dispute/src/RStarM.sol:250]: Still not quite right. See comment with permalink.

**Review by l-monninger — CHANGES_REQUESTED** [review]

**Comment by l-monninger** [comment]
@0xmovses these changes are mine to make.

**Comment by 0xmovses** [comment]
Do we still want RStarM on main?

**Comment by 0xmovses** [comment]
RStarM is not being shipped for now

**Pull request closed without merging**
